File: androidApp/src/main/kotlin/com/jrr/jrrkmp_native_ui/presentation/shell/LargeScreenShell.kt
package com.jrr.jrrkmp_native_ui.presentation.shell

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.List
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.SkipNext
import androidx.compose.material.icons.filled.SkipPrevious
import androidx.compose.material.icons.outlined.Pause
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.jrr.jrrkmp_native_ui.core.theme.AppColors
import com.jrr.jrrkmp_native_ui.core.theme.AppTypography
import com.jrr.jrrkmp_native_ui.presentation.components.VuMeter
import com.jrr.jrrkmp_native_ui.presentation.navigation.RootConfig
import coil.compose.AsyncImage

private val SIDEBAR_WIDTH = 256.dp

/**
 * Large-screen (tablet) chrome: a persistent left sidebar (brand, nav, docked
 * now-playing cell, active-zone chip) beside the routed [content]. Mirrors the
 * `Tablet & Desktop` design handoff. Navigation stays driven by [RootConfig] via
 * [onSelectTab]; this composable is presentation only.
 */
@Composable
fun LargeScreenShell(
    active: RootConfig,
    onSelectTab: (RootConfig) -> Unit,
    trackName: String?,
    trackArtist: String?,
    trackImageUrl: String?,
    isPlaying: Boolean,
    progress: Float,
    activeZoneName: String,
    onPlayPauseClick: () -> Unit,
    onNextClick: () -> Unit,
    onPrevClick: () -> Unit,
    content: @Composable () -> Unit,
) {
    Row(modifier = Modifier.fillMaxSize().background(AppColors.bg1)) {
        Sidebar(
            active = active,
            onSelectTab = onSelectTab,
            trackName = trackName,
            trackArtist = trackArtist,
            trackImageUrl = trackImageUrl,
            isPlaying = isPlaying,
            progress = progress,
            activeZoneName = activeZoneName,
            onPlayPauseClick = onPlayPauseClick,
            onNextClick = onNextClick,
            onPrevClick = onPrevClick,
        )
        Box(modifier = Modifier.weight(1f).fillMaxHeight().background(AppColors.bg1)) {
            content()
        }
    }
}

@Composable
private fun Sidebar(
    active: RootConfig,
    onSelectTab: (RootConfig) -> Unit,
    trackName: String?,
    trackArtist: String?,
    trackImageUrl: String?,
    isPlaying: Boolean,
    progress: Float,
    activeZoneName: String,
    onPlayPauseClick: () -> Unit,
    onNextClick: () -> Unit,
    onPrevClick: () -> Unit,
) {
    Column(
        modifier = Modifier
            .width(SIDEBAR_WIDTH)
            .fillMaxHeight()
            .background(AppColors.bg2),
    ) {
        // Brand
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(start = 22.dp, end = 22.dp, top = 22.dp, bottom = 18.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(34.dp)
                    .clip(CircleShape)
                    .background(AppColors.accent),
            )
            Spacer(Modifier.width(12.dp))
            Column {
                Text(
                    text = "JRIVER",
                    style = AppTypography.monoLabel.copy(
                        color = AppColors.text,
                        fontSize = 12.5.sp,
                        fontWeight = FontWeight.Medium,
                        letterSpacing = 2.sp,
                    ),
                )
                Text(
                    text = "REMOTE",
                    style = AppTypography.monoLabel.copy(
                        color = AppColors.accent,
                        fontSize = 9.sp,
                        letterSpacing = 2.sp,
                    ),
                )
            }
        }

        // Nav
        Column(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            NavItem("Now Playing", Icons.Default.PlayArrow, active == RootConfig.Player) { onSelectTab(RootConfig.Player) }
            NavItem("Library", Icons.Default.Home, active == RootConfig.Library) { onSelectTab(RootConfig.Library) }
            NavItem("Zones", Icons.Default.List, active == RootConfig.Zones) { onSelectTab(RootConfig.Zones) }
            NavItem("Settings", Icons.Default.Settings, active == RootConfig.Settings) { onSelectTab(RootConfig.Settings) }
        }

        Spacer(Modifier.weight(1f))

        // Docked now-playing cell
        if (!trackName.isNullOrEmpty()) {
            DockedNowPlaying(
                trackName = trackName,
                trackArtist = trackArtist ?: "",
                trackImageUrl = trackImageUrl,
                isPlaying = isPlaying,
                progress = progress,
                onBodyClick = { onSelectTab(RootConfig.Player) },
                onPlayPauseClick = onPlayPauseClick,
                onNextClick = onNextClick,
                onPrevClick = onPrevClick,
            )
        }

        // Active zone chip
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .padding(start = 12.dp, end = 12.dp, bottom = 14.dp)
                .fillMaxWidth()
                .clip(RoundedCornerShape(10.dp))
                .border(1.dp, AppColors.line, RoundedCornerShape(10.dp))
                .clickable { onSelectTab(RootConfig.Zones) }
                .padding(horizontal = 14.dp, vertical = 11.dp),
        ) {
            Box(modifier = Modifier.size(7.dp).clip(CircleShape).background(AppColors.success))
            Spacer(Modifier.width(10.dp))
            Column {
                Text(
                    text = "ACTIVE ZONE",
                    style = AppTypography.monoLabel.copy(color = AppColors.text3, fontSize = 9.sp, letterSpacing = 1.6.sp),
                )
                Spacer(Modifier.height(2.dp))
                Text(
                    text = activeZoneName,
                    style = AppTypography.itemSubtitle.copy(color = AppColors.text, fontWeight = FontWeight.Medium),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

@Composable
private fun NavItem(
    label: String,
    icon: ImageVector,
    selected: Boolean,
    onClick: () -> Unit,
) {
    val color = if (selected) AppColors.accent else AppColors.text3
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .height(46.dp)
            .clip(RoundedCornerShape(10.dp))
            .background(if (selected) AppColors.accentDim else androidx.compose.ui.graphics.Color.Transparent)
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp),
    ) {
        androidx.compose.material3.Icon(icon, contentDescription = label, tint = color, modifier = Modifier.size(20.dp))
        Spacer(Modifier.width(14.dp))
        Text(
            text = label,
            style = AppTypography.itemTitle.copy(color = color, fontSize = 14.5.sp, fontWeight = FontWeight.Medium),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
private fun DockedNowPlaying(
    trackName: String,
    trackArtist: String,
    trackImageUrl: String?,
    isPlaying: Boolean,
    progress: Float,
    onBodyClick: () -> Unit,
    onPlayPauseClick: () -> Unit,
    onNextClick: () -> Unit,
    onPrevClick: () -> Unit,
) {
    Column(
        modifier = Modifier
            .padding(12.dp)
            .fillMaxWidth()
            .clip(RoundedCornerShape(10.dp))
            .background(AppColors.bg3)
            .border(1.dp, AppColors.line2, RoundedCornerShape(10.dp))
            .clickable(onClick = onBodyClick),
    ) {
        // Progress hairline
        Box(
            modifier = Modifier
                .fillMaxWidth(progress.coerceIn(0f, 1f))
                .height(2.dp)
                .background(AppColors.accent),
        )
        Column(modifier = Modifier.padding(12.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Box(
                    modifier = Modifier
                        .size(44.dp)
                        .clip(RoundedCornerShape(4.dp))
                        .background(AppColors.bg4),
                ) {
                    if (!trackImageUrl.isNullOrEmpty()) {
                        AsyncImage(
                            model = trackImageUrl,
                            contentDescription = null,
                            modifier = Modifier.fillMaxSize(),
                        )
                    }
                }
                Spacer(Modifier.width(10.dp))
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = trackName,
                        style = AppTypography.itemSubtitle.copy(color = AppColors.text, fontWeight = FontWeight.Medium, fontSize = 13.sp),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                    Spacer(Modifier.height(2.dp))
                    Text(
                        text = trackArtist,
                        style = AppTypography.itemSubtitle.copy(color = AppColors.text2, fontSize = 11.5.sp),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
            ) {
                VuMeter(isPlaying = isPlaying)
                Row(verticalAlignment = Alignment.CenterVertically) {
                    CtlButton(Icons.Default.SkipPrevious, AppColors.text2, 18.dp, onPrevClick)
                    CtlButton(
                        if (isPlaying) Icons.Outlined.Pause else Icons.Default.PlayArrow,
                        AppColors.accent,
                        20.dp,
                        onPlayPauseClick,
                    )
                    CtlButton(Icons.Default.SkipNext, AppColors.text2, 18.dp, onNextClick)
                }
            }
        }
    }
}

@Composable
private fun CtlButton(icon: ImageVector, tint: androidx.compose.ui.graphics.Color, size: androidx.compose.ui.unit.Dp, onClick: () -> Unit) {
    Box(
        modifier = Modifier
            .size(32.dp)
            .clip(CircleShape)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        androidx.compose.material3.Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(size))
    }
}

File: iosApp/iosApp/Presentation/Views/Large/LargeScreenShell.swift
import SharedLogic
import SwiftUI

/// Large-screen (iPad regular width) chrome: a persistent left sidebar (brand,
/// nav, docked now-playing cell, active-zone chip) beside the routed `content`.
/// Mirrors the `Tablet & Desktop` design handoff. Navigation stays driven by
/// `RootConfig`; this view is presentation only.
struct LargeScreenShell<Content: View>: View {
    let activeTag: Int
    let onSelect: (Int) -> Void
    var nowPlaying: NowPlayingObservable
    @ViewBuilder let content: () -> Content

    /// Holds keyboard focus so the shell receives hardware-keyboard shortcuts.
    /// A focused text field (the library filter) takes focus on tap and
    /// consumes character keys first, so typing is never hijacked.
    @FocusState private var keyboardFocused: Bool

    private static var sidebarWidth: CGFloat {
        256
    }

    var body: some View {
        HStack(spacing: 0) {
            Sidebar(activeTag: activeTag, onSelect: onSelect, nowPlaying: nowPlaying)
                .frame(width: Self.sidebarWidth)
            Rectangle().fill(Color.line).frame(width: 1).ignoresSafeArea()
            content()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .background(Color.bg1)
        }
        .background(Color.bg1)
        .focusable()
        .focused($keyboardFocused)
        .onAppear { keyboardFocused = true }
        .onKeyPress { press in handleKey(press) }
    }

    private func handleKey(_ press: KeyPress) -> KeyPress.Result {
        switch press.key {
        case .space:
            if nowPlaying.isPlaying { nowPlaying.pause() } else { nowPlaying.play() }
            return .handled
        case .leftArrow:
            nowPlaying.previous(); return .handled
        case .rightArrow:
            nowPlaying.next(); return .handled
        case .upArrow:
            nowPlaying.setVolume(level: min(1, nowPlaying.volume + 0.05)); return .handled
        case .downArrow:
            nowPlaying.setVolume(level: max(0, nowPlaying.volume - 0.05)); return .handled
        case KeyEquivalent("q"):
            onSelect(2); return .handled
        case KeyEquivalent("l"):
            onSelect(0); return .handled
        default:
            return .ignored
        }
    }
}

private struct Sidebar: View {
    let activeTag: Int
    let onSelect: (Int) -> Void
    var nowPlaying: NowPlayingObservable

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Brand
            HStack(spacing: 12) {
                Circle().fill(Color.accentColor).frame(width: 34, height: 34)
                VStack(alignment: .leading, spacing: 2) {
                    Text("JRIVER")
                        .font(AppFont.ibmPlexMono(size: 12.5, weight: .medium))
                        .tracking(2)
                        .foregroundColor(.textPrimary)
                    Text("REMOTE")
                        .font(AppFont.ibmPlexMono(size: 9, weight: .regular))
                        .tracking(2)
                        .foregroundColor(.accentColor)
                }
            }
            .padding(.horizontal, 22)
            .padding(.top, 22)
            .padding(.bottom, 18)

            // Nav
            NavSection(activeTag: activeTag, onSelect: onSelect)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)

            Spacer()

            if nowPlaying.trackTitle != "Idle" {
                DockedNowPlaying(nowPlaying: nowPlaying, onBodyClick: { onSelect(2) })
                    .padding(12)
            }

            // Active zone chip
            Button(action: { onSelect(3) }) {
                HStack(spacing: 10) {
                    Circle().fill(Color.successColor).frame(width: 7, height: 7)
                    VStack(alignment: .leading, spacing: 2) {
                        Text("ACTIVE ZONE")
                            .font(AppFont.ibmPlexMono(size: 9, weight: .regular))
                            .tracking(1.6)
                            .foregroundColor(.textTertiary)
                        Text(nowPlaying.activeZoneName)
                            .font(AppFont.inter(size: 13, weight: .medium))
                            .foregroundColor(.textPrimary)
                            .lineLimit(1)
                    }
                    Spacer()
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 11)
                .overlay(RoundedRectangle(cornerRadius: 10).stroke(Color.line, lineWidth: 1))
            }
            .buttonStyle(.plain)
            .padding(.horizontal, 12)
            .padding(.bottom, 14)
        }
        .frame(maxHeight: .infinity, alignment: .top)
        .background(Color.bg2)
    }
}

/// Sidebar navigation list. On iOS 26 / macOS 26 the selected item's accent
/// glass pill *morphs* between rows as the selection changes: every item shares
/// one `glassEffectID` inside a `GlassEffectContainer`, so SwiftUI treats the
/// pill as a single glass entity and slides/stretches it across the transition
/// (taps in `ContentView` are wrapped in `withAnimation`). On older OSes it
/// falls back to the previous flat `accentDim` fill on the selected row.
private struct NavSection: View {
    let activeTag: Int
    let onSelect: (Int) -> Void

    /// Tag drives both selection and the upstream `RootConfig` mapping.
    private static let items: [(tag: Int, label: String, icon: String)] = [
        (2, "Now Playing", "play.circle.fill"),
        (0, "Library", "music.note.house.fill"),
        (3, "Zones", "speaker.wave.3.fill"),
        (4, "Settings", "gearshape.fill"),
    ]

    @Namespace private var glassNS

    var body: some View {
        VStack(spacing: 2) {
            ForEach(Self.items, id: \.tag) { item in
                NavItem(label: item.label, systemImage: item.icon, selected: activeTag == item.tag) {
                    onSelect(item.tag)
                }
                .slidingGlassPill(
                    selected: activeTag == item.tag,
                    id: "navSelection",
                    in: glassNS,
                    shape: RoundedRectangle(cornerRadius: 10, style: .continuous),
                )
            }
        }
        // Drive the pill slide implicitly off the selected tag. Selection
        // arrives asynchronously (Decompose → observable), so a `withAnimation`
        // around the tap can't capture it; an `.animation(_:value:)` here opens
        // the transaction whenever `activeTag` actually changes, so the
        // matched-geometry pill glides between rows.
        .animation(.spring(response: 0.42, dampingFraction: 0.8), value: activeTag)
    }
}

private struct NavItem: View {
    let label: String
    let systemImage: String
    let selected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 14) {
                Image(systemName: systemImage)
                    .font(.system(size: 18))
                    .frame(width: 20, height: 20)
                Text(label)
                    .font(AppFont.inter(size: 14.5, weight: .medium))
                    .lineLimit(1)
                Spacer()
            }
            .foregroundColor(selected ? .accentColor : .textTertiary)
            .padding(.horizontal, 14)
            .frame(height: 46)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

private struct DockedNowPlaying: View {
    var nowPlaying: NowPlayingObservable
    let onBodyClick: () -> Void

    private var progress: Double {
        nowPlaying.durationMs > 0 ? Double(nowPlaying.positionMs) / Double(nowPlaying.durationMs) : 0
    }

    var body: some View {
        VStack(spacing: 0) {
            // Progress hairline
            GeometryReader { geo in
                Rectangle()
                    .fill(Color.accentColor)
                    .frame(width: geo.size.width * CGFloat(min(max(progress, 0), 1)), height: 2)
            }
            .frame(height: 2)

            VStack(spacing: 12) {
                HStack(spacing: 10) {
                    JrrAsyncImage(
                        url: nowPlaying.imageUrl.isEmpty ? nil : URL(string: nowPlaying.imageUrl),
                        content: { image in image.resizable().aspectRatio(contentMode: .fill) },
                        placeholder: { Color.bg4 },
                    )
                    .frame(width: 44, height: 44)
                    .clipShape(RoundedRectangle(cornerRadius: 4))

                    VStack(alignment: .leading, spacing: 2) {
                        Text(nowPlaying.trackTitle)
                            .font(AppFont.inter(size: 13, weight: .medium))
                            .foregroundColor(.textPrimary)
                            .lineLimit(1)
                        Text(nowPlaying.artistName)
                            .font(AppFont.inter(size: 11.5, weight: .regular))
                            .foregroundColor(.textSecondary)
                            .lineLimit(1)
                    }
                    Spacer()
                }

                HStack {
                    VuMeter(isPlaying: nowPlaying.isPlaying)
                    Spacer()
                    HStack(spacing: 4) {
                        ctlButton("backward.fill", color: .textSecondary, size: 16) { nowPlaying.previous() }
                        ctlButton(nowPlaying.isPlaying ? "pause.fill" : "play.fill", color: .accentColor, size: 18) {
                            if nowPlaying.isPlaying { nowPlaying.pause() } else { nowPlaying.play() }
                        }
                        ctlButton("forward.fill", color: .textSecondary, size: 16) { nowPlaying.next() }
                    }
                }
            }
            .padding(12)
        }
        .liquidGlass(in: RoundedRectangle(cornerRadius: 10, style: .continuous))
        .contentShape(Rectangle())
        .onTapGesture(perform: onBodyClick)
    }

    private func ctlButton(_ systemName: String, color: Color, size: CGFloat, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: systemName)
                .font(.system(size: size))
                .foregroundColor(color)
                .frame(width: 32, height: 32)
        }
        .buttonStyle(.plain)
    }
}
